Flaky test: TestServerCredsDispatch

<details><summary>Log data</summary><p>

```
=== RUN   TestServerCredsDispatch
==================
WARNING: DATA RACE
Write by goroutine 177:
  google.golang.org/grpc/test.(*serverDispatchCred).ServerHandshake()
      /home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:4161 +0x5b
  google.golang.org/grpc.(*Server).useTransportAuthenticator()
      /home/travis/gopath/src/google.golang.org/grpc/server.go:431 +0x128
  google.golang.org/grpc.(*Server).handleRawConn()
      /home/travis/gopath/src/google.golang.org/grpc/server.go:502 +0x5a

Previous write by goroutine 183:
  google.golang.org/grpc/test.(*serverDispatchCred).ServerHandshake()
      /home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:4161 +0x5b
  google.golang.org/grpc.(*Server).useTransportAuthenticator()
      /home/travis/gopath/src/google.golang.org/grpc/server.go:431 +0x128
  google.golang.org/grpc.(*Server).handleRawConn()
      /home/travis/gopath/src/google.golang.org/grpc/server.go:502 +0x5a

Goroutine 177 (running) created at:
  google.golang.org/grpc.(*Server).Serve()
      /home/travis/gopath/src/google.golang.org/grpc/server.go:495 +0x8a7

Goroutine 183 (finished) created at:
  google.golang.org/grpc.(*Server).Serve()
      /home/travis/gopath/src/google.golang.org/grpc/server.go:495 +0x8a7

==================

panic: close of closed channel

goroutine 34954 [running]:
panic(0xab88c0, 0xc820230650)
	/home/travis/.gimme/versions/go1.6.4.linux.amd64/src/runtime/panic.go:481 +0x3ff
google.golang.org/grpc/test.(*serverDispatchCred).ServerHandshake(0xc820358700, 0x2adc57acb998, 0xc8203aa0a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/test/end2end_test.go:4162 +0x9b
google.golang.org/grpc.(*Server).useTransportAuthenticator(0xc8202652c0, 0x2adc57acb998, 0xc8203aa0a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/server.go:431 +0x129
google.golang.org/grpc.(*Server).handleRawConn(0xc8202652c0, 0x2adc57acb998, 0xc8203aa0a0)
	/home/travis/gopath/src/google.golang.org/grpc/server.go:502 +0x5b
created by google.golang.org/grpc.(*Server).Serve
	/home/travis/gopath/src/google.golang.org/grpc/server.go:495 +0x8a8
FAIL
```
</details>

--- a/stats/stats_test.go
+++ b/stats/stats_test.go
@@ -813,7 +813,9 @@ func testServerStats(t *testing.T, tc *testConfig, cc *rpcConfig, checkFuncs []f
 		err:         err,
 	}
 
+	h.mu.Lock()
 	checkConnStats(t, h.gotConn)
+	h.mu.Unlock()
 	checkServerStats(t, h.gotRPC, expect, checkFuncs)
 }
 

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -4145,21 +4145,22 @@ func TestNonFailFastRPCSucceedOnTimeoutCreds(t *testing.T) {
 }
 
 type serverDispatchCred struct {
-	ready   chan struct{}
-	rawConn net.Conn
+	rawConnCh chan net.Conn
 }
 
 func newServerDispatchCred() *serverDispatchCred {
 	return &serverDispatchCred{
-		ready: make(chan struct{}),
+		rawConnCh: make(chan net.Conn, 1),
 	}
 }
 func (c *serverDispatchCred) ClientHandshake(ctx context.Context, addr string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
 	return rawConn, nil, nil
 }
 func (c *serverDispatchCred) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
-	c.rawConn = rawConn
-	close(c.ready)
+	select {
+	case c.rawConnCh <- rawConn:
+	default:
+	}
 	return nil, nil, credentials.ErrConnDispatched
 }
 func (c *serverDispatchCred) Info() credentials.ProtocolInfo {
@@ -4172,8 +4173,7 @@ func (c *serverDispatchCred) OverrideServerName(s string) error {
 	return nil
 }
 func (c *serverDispatchCred) getRawConn() net.Conn {
-	<-c.ready
-	return c.rawConn
+	return <-c.rawConnCh
 }
 
 func TestServerCredsDispatch(t *testing.T) {
@@ -4192,8 +4192,10 @@ func TestServerCredsDispatch(t *testing.T) {
 	}
 	defer cc.Close()
 
+	rawConn := cred.getRawConn()
+	time.Sleep(100 * time.Millisecond)
 	// Check rawConn is not closed.
-	if n, err := cred.getRawConn().Write([]byte{0}); n <= 0 || err != nil {
+	if n, err := rawConn.Write([]byte{0}); n <= 0 || err != nil {
 		t.Errorf("Read() = %v, %v; want n>0, <nil>", n, err)
 	}
 }

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -4193,6 +4193,8 @@ func TestServerCredsDispatch(t *testing.T) {
 	defer cc.Close()
 
 	rawConn := cred.getRawConn()
+	// Give grpc a chance to see the error and potentially close the connection.
+	// And check that connection is not closed after that.
 	time.Sleep(100 * time.Millisecond)
 	// Check rawConn is not closed.
 	if n, err := rawConn.Write([]byte{0}); n <= 0 || err != nil {

